router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala

```diff
@@ -369,38 +391,50 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
         state match {
           case Closed(_) => false
           case state@RemoteClosed() =>
-            if (resetRemote(state, Reset.Closed)) false
+            if (resetFromLocal(state, Reset.Closed)) false
             else recv(hdrs)
 
           case RemoteOpen(remote@RemoteStreaming()) =>
-            if (resetRemote(remote, Reset.InternalError)) false
+            if (resetFromLocal(remote, Reset.InternalError)) false
             else recv(hdrs)
 
           case Open(remote@RemotePending()) =>
             val q = new AsyncQueue[Frame]
-            if (stateRef.compareAndSet(state, Open(RemoteStreaming(q)))) {
-              remote.setMessage(hdrs.headers, Stream(q))
-              true
-            } else recv(hdrs)
+            val msg = mkRecvMsg(hdrs.headers, Stream(q))
+            if (ConnectionHeaders.detect(msg.headers)) {
+              if (resetFromLocal(remote, Reset.ProtocolError)) true
+              else recv(hdrs)
+            } else {
+              if (stateRef.compareAndSet(state, Open(RemoteStreaming(q)))) {
+                remote.setMessage(msg)
+                true
+              } else recv(hdrs)
+            }
 
           case LocalClosed(remote@RemotePending()) =>
             val q = new AsyncQueue[Frame]
-            if (stateRef.compareAndSet(state, LocalClosed(RemoteStreaming(q)))) {
-              remote.setMessage(hdrs.headers, Stream(q))
-              true
-            } else recv(hdrs)
+            val msg = mkRecvMsg(hdrs.headers, Stream(q))
+            if (ConnectionHeaders.detect(msg.headers)) {
```

Comment:
I gave it a shot but it's a little tricky to do without allocating. going to punt for now